### ISSUE TITLE
llvm.amdgcn.buffer.atomic.fadd.f32 Intrinsic has incorrect return type

### ISSUE THREAD

**Issue by jeffdaily** [ROCmSoftwarePlatform/MIOpen#650]
Built a PyTorch wheel based on PyTorch 1.7 branch and MIOpen develop branch commit 81a206f57dad1fb3620fc55e354fc2b8833647a5.

```
commit 81a206f57dad1fb3620fc55e354fc2b8833647a5
Author: Artem Tamazov <artem.tamazov@gmail.com>
Date:   Mon Dec 7 19:39:26 2020 +0300

    Workaround for SWDEV-262823 (#633)
```

When executing the pytorch microbenchmark, the following error occurred at runtime.

```
Intrinsic has incorrect return type!
float (float, <4 x i32>, i32, i32, i1)* @llvm.amdgcn.buffer.atomic.fadd.f32

MIOpen Error: /MIOpen/src/hipoc/hipoc_program.cpp:285: Code object build failed. Source: gridwise_convolution_backward_weights_implicit_gemm_v4r4_xdlops_nchw_kcyx_nkhw.cpp
```

**Comment by sunway513** [comment]
Some more logs in case helpful:
```
MIOpen(HIP): Info2 [LoadBinary] Unable to load binary for: gridwise_convolution_backward_weights_implicit_gemm_v4r4_xdlops_nchw_kcyx_nkhw.cpp ;args
:  -DCK_PARAM_PROBLEM_G=1 -DCK_PARAM_PROBLEM_N=256 -DCK_PARAM_PROBLEM_K=2048 -DCK_PARAM_PROBLEM_C=512 -DCK_PARAM_PROBLEM_HI=7 -DCK_PARAM_PROBLEM_WI
=7 -DCK_PARAM_PROBLEM_HO=7 -DCK_PARAM_PROBLEM_WO=7 -DCK_PARAM_PROBLEM_Y=1 -DCK_PARAM_PROBLEM_X=1 -DCK_PARAM_PROBLEM_CONV_STRIDE_H=1 -DCK_PARAM_PROB
LEM_CONV_STRIDE_W=1 -DCK_PARAM_PROBLEM_CONV_DILATION_H=1 -DCK_PARAM_PROBLEM_CONV_DILATION_W=1 -DCK_PARAM_PROBLEM_IN_LEFT_PAD_H=0 -DCK_PARAM_PROBLEM
_IN_LEFT_PAD_W=0 -DCK_PARAM_PROBLEM_IN_RIGHT_PAD_H=0 -DCK_PARAM_PROBLEM_IN_RIGHT_PAD_W=0 -DCK_PARAM_TUNABLE_GEMM_M_PER_BLOCK=128 -DCK_PARAM_TUNABLE
_GEMM_N_PER_BLOCK=128 -DCK_PARAM_TUNABLE_GEMM_K_PER_BLOCK=4 -DCK_PARAM_TUNABLE_GEMM_M_PER_WAVE=64 -DCK_PARAM_TUNABLE_GEMM_N_PER_WAVE=64 -DCK_PARAM_
TUNABLE_GEMM_KPACK=8 -DCK_PARAM_GEMM_K_BLOCK=8 -DCK_PARAM_DEPENDENT_BLOCK_SIZE=256 -DCK_PARAM_DEPENDENT_GRID_SIZE=512 -DCK_PARAM_DEPENDENT_GEMM_A_B
LOCK_COPY_CLUSTER_LENGTHS_GEMM_K=4 -DCK_PARAM_DEPENDENT_GEMM_A_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_M=8 -DCK_PARAM_DEPENDENT_GEMM_A_BLOCK_COPY_CLUSTER_L
ENGTHS_GEMM_KPACK=8 -DCK_PARAM_DEPENDENT_GEMM_A_BLOCK_COPY_SRC_DATA_PER_READ_GEMM_KPACK=1 -DCK_PARAM_DEPENDENT_GEMM_A_BLOCK_COPY_DST_DATA_PER_WRITE
_GEMM_KPACK=1 -DCK_PARAM_DEPENDENT_GEMM_B_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_K=4 -DCK_PARAM_DEPENDENT_GEMM_B_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_N=8 -DCK_
PARAM_DEPENDENT_GEMM_B_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_KPACK=8 -DCK_PARAM_DEPENDENT_GEMM_B_BLOCK_COPY_SRC_DATA_PER_READ_GEMM_KPACK=1 -DCK_PARAM_DEP
ENDENT_GEMM_B_BLOCK_COPY_DST_DATA_PER_WRITE_GEMM_KPACK=1 -DCK_USE_AMD_XDLOPS=1 -DCK_USE_AMD_XDLOPS_INLINE_ASM=0 -DCK_USE_AMD_XDLOPS_EMULATE=0 --std
=c++14 -DCK_USE_AMD_BUFFER_ATOMIC_FADD=1 -DCK_BLOCK_SYNC_LDS_WITHOUT_SYNC_VMEM=1 -DCK_WORKAROUND_SWDEV_229564=1 -DCK_WORKAROUND_SWDEV_231101=1 -DCK
_USE_AMD_BUFFER_ADDRESSING=1 -DCK_USE_AMD_V_FMAC_F32=0 -DMIOPEN_USE_FP16=1 -DMIOPEN_USE_FP32=0 -DMIOPEN_USE_INT8=0 -DMIOPEN_USE_INT8x4=0 -DMIOPEN_U
SE_BFP16=0 -DMIOPEN_USE_INT32=0 -DMIOPEN_USE_RNE_BFLOAT16=1 -mcpu=gfx908                                                                           
MIOpen(HIP): Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx908+sram-ecc                                                                                 
MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_LINK_BC_TO_BC: ERROR (1)            
MIOpen(HIP): Error [BuildHip] comgr status = ERROR (1)                                                                                             
MIOpen(HIP): Warning [BuildHip] Intrinsic has incorrect return type!                                                                               
float (float, <4 x i32>, i32, i32, i1)* @llvm.amdgcn.buffer.atomic.fadd.f32                                                                        Intrinsic has incorrect return type!   
```

**Comment by atamazov** [comment]
@jeffdaily Are you intentionally use COMgr?

**Comment by atamazov** [comment]
@jeffdaily What is the version of HIP? The library reports it in the following log line (once):
```
Info [HipCompilerVersionImpl] x.y.z
```

The `llvm.amdgcn.buffer.atomic.fadd.f32` intrinsic had a `void` return type before 3.10. Since ROCm-3.10, the return type becomes `float`. What is valid return type now?

@asroy The fix should be provided here:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/fa04cffdbb8550452cfd7ccdaeec2b9039770b57/src/kernels/composable_kernel/include/utility/amd_buffer_addressing.hpp#L142-L153

**Comment by atamazov** [comment]
> The `llvm.amdgcn.buffer.atomic.fadd.f32` intrinsic had a `void` return type before 3.10. Since ROCm-3.10, the return type becomes `float`. What is valid return type now?

Perhaps David Salinas can answer, [see here](http://ontrack-internal.amd.com/browse/SWDEV-255582?focusedCommentId=6465612&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-6465612).

**Comment by atamazov** [comment]
@jeffdaily And thanks for opening an issue here.

**Comment by zjing14** [comment]
@jeffdaily @sunway513 What is the rocm version/docker you are using? I cannot reproduce the issue.

**Comment by atamazov** [comment]
@zjing14 I am reaching them by mail

**Comment by atamazov** [comment]
@jeffdaily please look at HIP version. The library reports it in the following log line (only once):
```
Info [HipCompilerVersionImpl] x.y.z
```
If you're using ROCm 3.10, then it should be >= 3.10.20405.

**Comment by atamazov** [comment]
We need to check if ROCm 3.10 really wants "void" intrinsic and then decide about a fix.

**Comment by daniellowell** [comment]
@jeffdaily A number of questions have been asked here, can you provide details?

**Comment by asroy** [comment]
@jeffdaily from the email, you are using 3.10.20465, which doesn't have the intrinsic return type change yet, resulting compilation failure.
```
MIOpen(HIP): Info [HipCompilerVersionImpl] 3.10.20465
```

However, QA's docker 3.10.20412-d9f655f5 already contains the type change. This is strange.
http://ontrack-internal.amd.com/browse/SWDEV-255582?focusedCommentId=6549548&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-6549548I

I created PR #671, but I don't think it's safe to merge into MIOpen develop branch.

cc @atamazov, @daniellowell

**Comment by asroy** [comment]
According to compiler team, the compiler change in this JIRA (http://ontrack-internal.amd.com/browse/SWDEV-255582) has NOT made into rocm3.10 or rocm4.0 as planned. Therefore, the return type of buffer atomic in rocm3.10 and 4.0 should be still be ```void``` and NOT ```float```


This issue reported by @jeffdaily is reproducible on gfx908 with release docker ```rocm/pytorch:rocm4.0_ubuntu18.04_py3.6_pytorch```. But you need to build MIOpen on HIP backend, and turn on  COMGR in order to reproduce it
```
cmake                                                                                                     \
-DMIOPEN_BACKEND=HIP                                                                                      \
-DCMAKE_CXX_COMPILER=/opt/rocm/llvm/bin/clang++                                                           \
-DMIOPEN_USE_COMGR=ON                                                                                     \
```

and using the cmd 
```
 export MIOPEN_DEBUG_CONV_WINOGRAD=0
 export MIOPEN_DEBUG_CONV_FFT=0
 export MIOPEN_DEBUG_CONV_DIRECT=0
 export MIOPEN_DEBUG_CONV_GEMM=0
 export MIOPEN_DEBUG_CONV_SCGEMM=0
 export MIOPEN_DEBUG_CONV_IMPLICIT_GEMM=1
 export MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS=1
 bin/MIOpenDriver conv -F 4 -n 256 -g 1 -k  256 -c 1024 -H  14 -W  14 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -V 0 -w 1 -t 1 -i 10
```

Observation:
1. For MIOpen built with COMPGR on, compiler will complain about return type being ```float```, if you change return type to ```void```, compiler will no longer complain.
2. However, For MIOpen built with COMGPR off, compiler will NOT complain no matter if return type is ```float``` or ```void```. The reason why it takes both return type is unknown.

**Comment by atamazov** [comment]
@asroy Thank you for the investigations! :+1: 

> According to compiler team, the compiler change in this JIRA (http://ontrack-internal.amd.com/browse/SWDEV-255582) has NOT made into rocm3.10 or rocm4.0 as planned.

Some kind of confusion with HIP versions. How did it happen that everything worked fine in the staging with version 3.10.20412 (that was 3.10 RC), but then the new intrinsics did not get into either 3.10 or 4.0? I don't clearly understand what is going on.

> Therefore, the return type of buffer atomic in rocm3.10 and 4.0 should still be void and NOT float

The simplest hot-fix is changing 
```C++
 #if CK_HIP_VERSION_FLAT >= 3010020405 
 // starting ROCm-3.10, the return type becomes float 
```
to
```C++
 #if CK_HIP_VERSION_FLAT >= 99999999999
 // It is planned to change return type from void to float in some future version
```

**Comment by atamazov** [comment]
> According to compiler team, the compiler change in this JIRA (http://ontrack-internal.amd.com/browse/SWDEV-255582) has NOT made into rocm3.10 or rocm4.0 as planned.

But they have it in the staging? If so then, removal of `float` return type will lead to failure in the staging.

**Comment by asroy** [comment]
QA was using a 3.10.20412 rocm docker, but a newer compiler for the JIRA test.

**Comment by atamazov** [comment]
@asroy @daniellowell 
> QA was using a 3.10.20412 rocm docker, but with a newer compiler in it

This may mean complete mess. It is likely that 671 will break their staging. To keep their staging in workable state, we need detailed HIP compiler version information. What we have right now does not give us necessary information:

```bash
# /opt/rocm/llvm/bin/clang++ --version
clang version 12.0.0 (/src/external/llvm-project/clang 927e2776dc0e4bb0119efbc5ea405b7425d7f4ac)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
```

Alternatively, _we can add a CMake variable_ to tell MIOpen that HIP compiler and libraries are of some RC version, like this:
```
cmake ... -DMIOPEN_ENFORCE_HIP_PACKAGE_VERSION=MMMmmmpppppp
```
QA will need to set this CMake variable to something like 4001000000.
